model/base_code_model.py

```diff
@@ -0,0 +1,158 @@
+from abc import ABCMeta, abstractmethod
+from math import ceil
+from typing import Tuple, Dict, List, Union
+
+import torch
+from pytorch_lightning.core.lightning import LightningModule
+from torch.optim import Adam, Optimizer, SGD
+from torch.optim.lr_scheduler import ExponentialLR, _LRScheduler
+from torch.utils.data import DataLoader
+
+from configs import ModelHyperparameters
+from dataset import Vocabulary, create_dataloader, PathContextBatch
+from utils.metrics import SubtokenStatistic, ClassificationStatistic
+
+StatisticType = Union[SubtokenStatistic, ClassificationStatistic]
+
+
+class BaseCodeModel(LightningModule, metaclass=ABCMeta):
+    def __init__(
+        self, config: ModelHyperparameters, vocab: Vocabulary, num_workers: int = 0,
+    ):
+        super().__init__()
+        self.save_hyperparameters()
+        self.config = config
+        self.vocab = vocab
+        self.num_workers = num_workers
+
+    def forward(
```

Comment:
This is an abstract method, forward pass can be different for each model. Moreover, I don't sure, that code2class will use the same arguments, for example, it doesn't need target_sequnce for teacher forcing.